### PR TITLE
qemu chardev-argo fixes

### PR DESCRIPTION
**Pull request by jandryuk**
qemu can go into an infinite loop if it tries to send a message bigger than the destination's ring size or available space.  argo_send() returns EMSGSIZE & argo_chr_write() returns 0.  The chardev code thinks nothing was sent, so it calls back into argo_chr_write() and the process repeats with no progress made.  Change argo_chr_write() to scale back the write size when receiving EMSGSIZE.

Also use the Xen provided define to check for invalid domids.

### CODE REVIEW

**Comment by eric-ch** [comment]
Build [$6714](http://openxt-builder.ainfosec.com:8010/builders/openxt64/builds/6714)

**Comment by crogers1** [comment]
I'm interested in taking this for stable-9, if you could open a PR once this master one is approved, that would be very appreciated.

**Comment by stacktrust** [comment]
Please create a JIRA ticket for this fix and associated test case.

Do we need to document how non-QEMU users of chardev-argo can avoid a similar loop?

File: recipes-extended/qemu-dm/qemu-dm/qmp-argo-char-driver.patch [hunk 1]
```diff
@@ -85,20 +85,37 @@
...
++            ret = argo_sendto(s->fd, buf, write_len, 0, &s->remote_addr);
++        }
++
++        if ( ret == -1 && errno == EMSGSIZE) {
++            /* If the write was too big, try a smaller piece */
++            write_len /= 2;
++            continue;
++        } else if (ret == -1) {
++            fprintf(stderr, "%s error: argo_sendto() failed (%s) - %d %d.\n",
++                    ARGO_CHARDRV_NAME, strerror(errno), ret, len);
```
Comment (dozylynx) [inline recipes-extended/qemu-dm/qemu-dm/qmp-argo-char-driver.patch]: This PR has really good improvements to the code - I especially like the use of clear variable names and the logic to step down the size of the workload to the point that the receiver has available capacity for. Thanks for doing this.
The only (super minor) point I can see in this section is about the error message in the above line -- which was inherited from the prior code: it's not always the case that `argo_sendto` is the function that was invoked -- ie. `argo_send` is called instead for streams. Is the existing error message close enough? Up to you - the improvements here are worthwhile anyway.

**Comment by jandryuk** [comment]
Pushed new version based off the earlier comments with an updated commit message.

**Comment by eric-ch** [comment]
Build [$6731](http://openxt-builder.ainfosec.com:8010/builders/openxt64/builds/6731)

**Comment by jandryuk** [comment]
I force pushed with updated commit messages to capture a little more detail about the changes.

**Comment by eric-ch** [comment]
Build [$6742](http://openxt-builder.ainfosec.com:8010/builders/openxt64/builds/6742)

**Comment by eric-ch** [comment]
Will merge soon.

**Pull request merged**